File: src/lib/segment.ts
/**
 * pain ai — Sentence Segmentation Engine (segment.ts)
 * 
 * Splits incoming agent response text into discrete, spoken sentences for dual-render
 * TTS audio generation and real-time caption synchronization.
 * Guards against false splits on abbreviations, decimals, and quote boundaries.
 */

export interface CaptionSentenceItem {
  i: number;
  text: string;
}

const ABBREVIATIONS = new Set([
  'mr', 'mrs', 'ms', 'dr', 'prof', 'sr', 'jr', 'st', 'no', 'vs',
  'eg', 'ie', 'etc', 'am', 'pm', 'fig', 'al', 'dept', 'gen', 'gov',
  'jan', 'feb', 'mar', 'apr', 'jun', 'jul', 'aug', 'sep', 'sept', 'oct', 'nov', 'dec'
]);

const PUNCT_SPLIT_REGEX = /([.!?…]+["']?)(?:\s+|$)/g;

function isAbbreviation(prefix: string): boolean {
  const cleaned = prefix.trim().toLowerCase().replace(/\./g, '');
  if (!cleaned) return false;
  if (ABBREVIATIONS.has(cleaned)) return true;
  // Single letter initial (e.g. "J." in "J. Doe")
  if (cleaned.length === 1 && /[a-z]/i.test(cleaned)) return true;
  return false;
}

export function splitSentences(text: string): CaptionSentenceItem[] {
  if (!text || !text.trim()) {
    return [];
  }

  const rawText = text.trim();
  const placeholder = '\u2060'; // Word joiner non-printing character

  // Protect known abbreviations
  let protectedText = rawText.replace(/\b([a-zA-Z]{1,5})\./g, (match, word) => {
    if (isAbbreviation(word)) {
      return word + placeholder;
    }
    return match;
  });

  // Protect decimals: 3.14 -> 3\u206014
  protectedText = protectedText.replace(/(\d+)\.(\d+)/g, `$1${placeholder}$2`);

  // Split on sentence boundaries: [.!?…]+ followed by whitespace or end of string
  const parts: string[] = [];
  let lastIdx = 0;

  // Reset regex state
  PUNCT_SPLIT_REGEX.lastIndex = 0;

  while (PUNCT_SPLIT_REGEX.exec(protectedText) !== null) {
    const endPos = PUNCT_SPLIT_REGEX.lastIndex;
    const candidate = protectedText.substring(lastIdx, endPos).trim();
    if (candidate) {
      parts.push(candidate);
    }
    lastIdx = endPos;
  }

  if (lastIdx < protectedText.length) {
    const remainder = protectedText.substring(lastIdx).trim();
    if (remainder) {
      parts.push(remainder);
    }
  }

  // Restore placeholders and build sentence items
  const sentences: CaptionSentenceItem[] = [];
  let itemIdx = 0;

  for (const chunk of parts) {
    const restored = chunk.replaceAll(placeholder, '.').trim();
    if (restored) {
      sentences.push({
        i: itemIdx,
        text: restored,
      });
      itemIdx++;
    }
  }

  if (sentences.length === 0 && rawText) {
    sentences.push({ i: 0, text: rawText });
  }

  return sentences;
}

File: sidecar/voice/segment.py
#!/usr/bin/env python3
"""pain ai — Sentence Segmentation Engine (segment.py)

Splits text into discrete, spoken sentences for dual-render TTS audio generation
and real-time caption synchronization. Guards against false splits on common
abbreviations, decimal numbers, and closing quotation marks.
"""

import re
from typing import Any, Dict, List

# Common abbreviations that should not trigger a sentence break
ABBREVIATIONS = {
    "mr", "mrs", "ms", "dr", "prof", "sr", "jr", "st", "no", "vs",
    "eg", "ie", "etc", "am", "pm", "fig", "al", "dept", "gen", "gov",
    "jan", "feb", "mar", "apr", "jun", "jul", "aug", "sep", "sept", "oct", "nov", "dec"
}

# Regex for sentence terminators followed by whitespace or end of string
# Matches combinations of ., !, ?, …, optionally followed by closing quotes or brackets
PUNCT_SPLIT_REGEX = re.compile(r'([.!?…]+["\']?)(?:\s+|$)')


def _is_abbreviation(prefix: str) -> bool:
    """Checks if the word preceding a period is a known abbreviation or single initial."""
    cleaned = prefix.strip().lower().rstrip(".").replace(".", "")
    if not cleaned:
        return False
    if cleaned in ABBREVIATIONS:
        return True
    # Single uppercase initial (e.g. "J." in "J. Doe")
    if len(cleaned) == 1 and cleaned.isalpha():
        return True
    return False


def split_sentences(text: str) -> List[Dict[str, Any]]:
    """Splits text into a list of sentence items with 0-based indices.

    Returns:
        List of dicts: [{"i": 0, "text": "First sentence."}, ...]
    """
    if not text or not text.strip():
        return []

    raw_text = text.strip()

    # Step 1: Protect known abbreviations by replacing their period with a placeholder
    placeholder = "\u2060"  # Word joiner non-printing character

    def protect_match(m: re.Match) -> str:
        word = m.group(1)
        if _is_abbreviation(word):
            return word + placeholder
        return m.group(0)

    # Protect abbreviations like e.g. and i.e.
    protected = re.sub(r'\b([a-zA-Z]{1,5})\.', protect_match, raw_text)

    # Protect decimals: digit.digit
    protected = re.sub(r'(\d+)\.(\d+)', r'\1' + placeholder + r'\2', protected)

    # Step 2: Split on sentence boundaries: [.!?…]+ followed by space or newline
    parts = []
    last_idx = 0

    for m in PUNCT_SPLIT_REGEX.finditer(protected):
        end_pos = m.end()
        candidate = protected[last_idx:end_pos].strip()
        if candidate:
            parts.append(candidate)
        last_idx = end_pos

    # Capture any remainder text
    if last_idx < len(protected):
        remainder = protected[last_idx:].strip()
        if remainder:
            parts.append(remainder)

    # Step 3: Restore placeholders and assemble sentence list
    sentences: List[Dict[str, Any]] = []
    item_idx = 0

    for chunk in parts:
        restored = chunk.replace(placeholder, ".").strip()
        if restored:
            sentences.append({
                "i": item_idx,
                "text": restored
            })
            item_idx += 1

    # Fallback if no punctuation splitting occurred but text exists
    if not sentences and raw_text:
        sentences.append({"i": 0, "text": raw_text})

    return sentences


if __name__ == "__main__":
    import json
    import sys

    sample = sys.argv[1] if len(sys.argv) > 1 else "Hello world! Dr. Smith arrived at 3.14 p.m. Can we begin? Yes."
    result = split_sentences(sample)
    print(json.dumps(result, indent=2))
